affordance_primitives/include/affordance_primitives/screw_planning/screw_planning.hpp

```diff
@@ -36,22 +36,71 @@
-  const Eigen::Isometry3d & current_pose, const Eigen::Isometry3d & start_pose,
-  const ScrewAxis & screw_axis, double theta_max, double theta_guess,
-  Eigen::Ref<Eigen::VectorXd> out);
+/**
+* @brief Computes the set of screw angles that minimize the error between EE pose and screw path 
+*
+* @param tf_m_to_q Forward kinematics of the robot
+* @param tf_m_to_s Start pose
+* @param screw_axis_set Set of screw axes (S1,...,Sm) in frame M
+* @param phi_max Set of screw distances to move
+* @param phi_start Set of initial random guesses for screw angles
+*
+* @return Constraint function success, error, and corresponding screw angles
+*/
+std::tuple<bool, Eigen::VectorXd, Eigen::VectorXd> constraintFn(
+  const Eigen::Isometry3d & tf_m_to_q, const Eigen::Isometry3d & tf_m_to_s,
+  const std::vector<ScrewAxis> & screw_axis_set, const Eigen::VectorXd phi_max,
+  const Eigen::VectorXd phi_start);
+
+/**
+ * @brief Computes the partial derivative of error between EE pose and screw path
+ *
+ * @param tf_q_to_m Inverse of the forward kinematics matrix, tf_m_to_q
+ * @param tf_m_to_s Transform from base frame, M to screw path
+ * @param phi_current Current set of screw angles
+ * @param screw_axis_set Set of screw axes (S1,...,Sm) in frame M
+ *
+ * @return Vector of partial derivative of error with respect of screw angles
+ */
+Eigen::VectorXd errorDerivative(
+  const Eigen::Isometry3d & tf_q_to_m, const Eigen::Isometry3d & tf_m_to_s,
+  const Eigen::VectorXd & phi_current, const std::vector<ScrewAxis> & screw_axis_set);
 
-Eigen::VectorXd calcError(const Eigen::Isometry3d & tf_err);
+/**
+ * @brief Computes the 6x1 error vector for a given transformation matrix
+ *
+ * @param tf Transformation matrix
+ *
+ * @return 6x1 Error vector
+ */
+Eigen::VectorXd eta(const Eigen::Isometry3d & tf);
 
-double calcErrorDerivative(
-  const Eigen::Isometry3d & tf_m_to_q, const Eigen::Isometry3d & tf_m_to_e,
-  const double current_theta, const ScrewAxis & screw_axis);
+/**
+ * @brief Given a set of screw axes and angles, computes the product of exponentials for specified start and end indices
+ *
+ * @param screw_axis_set Set of screw axes (S1,...,Sm)
+ * @param phi Set of screw angles
+ * @param start Start index for the product
+ * @param end End index for the product
+ *
+ * @return Homogenous transformation matrix representing the product of exponentials
+ */
+Eigen::Isometry3d productOfExponentials(
+  const std::vector<ScrewAxis> & screw_axis_set, const Eigen::VectorXd & phi, int start, int end);
 
-std::pair<double, Eigen::Isometry3d> findClosestPoint(
-  const Eigen::Isometry3d & tf_m_to_q, const Eigen::Isometry3d & tf_m_to_e,
-  const double theta_start, const double theta_max, const ScrewAxis & screw_axis);
+/**
+ * @brief Clamps the values of array to given highs and lows 
+ *
+ * @param arr array that is to be clamped
+ * @param arr_low array of lows
+ * @param arr_high array of highs
+ *
+ * @return Clamped array 
+ */
+Eigen::VectorXd clamp(
```

Comment:
I'm already using std::clamp in here. However, as far as I can see, it only works to clamp a single number, not an array(or a vector) as we need here. Hence, I put it into a function that employs a loop to clamp an array. We don't need to expose it to a higher level user in this case, but I surmise it might be handy in the future if we move it to a more general file, maybe affordance_utils, IDK? Another option is to absorb it in the function that calls it.

Reply:
Thinking about this a bit, I think what we might do is move the declaration for this function, and maybe a few others here (error derivative, POE?) into the c++ file. They are more of helper functions, and I am not sure they need to be exposed to anybody for use outside of the internals of the ap planning here. What do you think?

Reply:
You are saying don't have them in the hpp, but just have them in the cpp?

Reply:
Yea. I cannot think of a case when another user should have to do like include this header and then use PoE or something. And if we did want to expose that, it should be outside the planning context and with the screw model context instead, right?

Reply:
Actually, I surmise I'd be reusing these functions after you are gone. pOE would be recurring whenever I'd use Lynch/ Alameigi's stuff. It could, maybe, belong to affordance_utils in screw model. I foresee reusing clamp as well at some point, but that is a very general math function, and I don't know which header it should belong to. errorDerivative, calculateEta, constraintFn may be reused if I develop on this work more and add additional constraints. We should, maybe, leave them in the screw_planning header.